src/Parser.php

```diff
@@ -363,15 +358,26 @@ protected function isIpv6Host(string $ipv6): bool
      */
     protected function isRegisteredName(string $host): bool
     {
+        static $pattern = '/^(?:[a-z0-9\-_]{1,63}\.){0,126}[a-z0-9\-_]{1,63}$/i';
```

Comment:
Not in a registered name, as far as I know you can use it in a DNS name, so subdomains could.

Reply:
`_` and `-` are allowed because i had to resolve [the following bug](https://github.com/thephpleague/uri/issues/107). A DNS name is a registered name but a registered name is not always a DNS name

Reply:
Right, but then also `~` and a few more characters (`sub-delims`) needs to be allowed and the length requirements as well as the IDNA normalization need to be entirely dropped. See https://tools.ietf.org/html/rfc3986#section-3.2.2 + https://tools.ietf.org/html/rfc3986#appendix-A.

>       reg-name    = *( unreserved / pct-encoded / sub-delims )

>    unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"

>   sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
                 / "*" / "+" / "," / ";" / "="